src/module_4/module_4_model-fitting.md

```diff
@@ -0,0 +1,1214 @@
+
+
+```python
+train_cols = numerical_cols + binary_cols
+
+n_trees_grid = [5, 25, 50, 100]
+
+metrics = {
+    'n_trees': [],
+    'train_auc': [],
+    'val_auc': [],
+    'train_ap': [],
+    'val_ap': [],
+    'train_ce': [],
+    'val_ce': []
+}
+
+for n_trees in n_trees_grid:
+    rf = RandomForestClassifier(n_estimators=n_trees)
+    rf.fit(X_train[train_cols], y_train)
+    
+    train_probs = rf.predict_proba(X_train[train_cols])[:, 1]
+    val_probs = rf.predict_proba(X_val[train_cols])[:, 1]
+    
+    metrics['n_trees'].append(n_trees)
+    metrics['train_auc'].append(roc_auc_score(y_train, train_probs))
+    metrics['val_auc'].append(roc_auc_score(y_val, val_probs))
+    metrics['train_ap'].append(average_precision_score(y_train, train_probs))
+    metrics['val_ap'].append(average_precision_score(y_val, val_probs))
+    metrics['train_ce'].append(log_loss(y_train, train_probs))
+    metrics['val_ce'].append(log_loss(y_val, val_probs))
+
+print("Métricas:")
+for i in range(len(n_trees_grid)):
+    print(f"# Trees: {metrics['n_trees'][i]}")
+    print(f"Train AUC: {metrics['train_auc'][i]:.4f}, Validation AUC: {metrics['val_auc'][i]:.4f}")
+    print(f"Train AP: {metrics['train_ap'][i]:.4f}, Validation AP: {metrics['val_ap'][i]:.4f}")
+    print(f"Train Cross Entropy: {metrics['train_ce'][i]:.4f}, Validation Cross Entropy: {metrics['val_ce'][i]:.4f}")
+    print()
+```
+
+    Métricas:
+    # Trees: 5
+    Train AUC: 0.9909, Validation AUC: 0.6442
+    Train AP: 0.7707, Validation AP: 0.0707
+    Train Cross Entropy: 0.0243, Validation Cross Entropy: 0.3497
+    
+    # Trees: 25
+    Train AUC: 0.9952, Validation AUC: 0.7161
+    Train AP: 0.8524, Validation AP: 0.1128
+    Train Cross Entropy: 0.0209, Validation Cross Entropy: 0.2489
+    
+    # Trees: 50
+    Train AUC: 0.9955, Validation AUC: 0.7302
+    Train AP: 0.8598, Validation AP: 0.1185
+    Train Cross Entropy: 0.0207, Validation Cross Entropy: 0.2231
+    
+    # Trees: 100
+    Train AUC: 0.9957, Validation AUC: 0.7484
+    Train AP: 0.8628, Validation AP: 0.1248
+    Train Cross Entropy: 0.0206, Validation Cross Entropy: 0.1915
+    
+
+
+Tras ejecutar este código, podemos observar los siguientes aspectos:
+
+- La elevada precisión obtenida en el conjunto de entrenamiento es un resultado esperado, considerando que estamos implementando un modelo de bosque aleatorio.
+- Se constata una mejora considerable en el rendimiento del modelo con el aumento en el número de árboles. Aunque sería factible continuar ajustando este parámetro, experimentando con valores como 150 o 200, las restricciones de tiempo nos llevan a establecer 100 como el número óptimo de árboles.
+
+### Model 1. Random Forest - Without categorical variables
+
+Iniciaremos el entrenamiento de un primer modelo utilizando los parámetros previamente seleccionados y enfocándonos únicamente en las variables numéricas. Por el momento, excluiremos las variables categóricas debido a su complejidad.
```

Comment:
What's the difference between this and the RF that you just fit before? you are choosing the best one? why?